[BB-6151] feat: upgrades tinyMCE v4.0.20 to tinyMCE v5.5.1

DubeySandeep opened the pull request:
## Description

This PR upgrades TinyMCE version 4 to version 5.5.1.

These changes will impact the "Course author".

## Supporting information
- [Discovery doc to upgrade TinyMCE **[Approved]**](https://docs.google.com/document/d/1glPBW6OYkhkn_jAJndC8JMyruHl4RY9YGd4bqWbySg0/edit)
- Internal ticket: https://tasks.opencraft.com/browse/BB-6151

## Testing instructions

#### Testing new TinyMCE in Studio
1. Run the studio server from devstack dir “make studio-up”
1. Login to studio and create a new course.
1. Add a new unit in the course and go to the unit editor page.
1. Click “HTML” component button and select “Text”.
1. Click the component to edit, this will show a modal containing “tinymce” editor.
1. Use the TinyMCE editor thoroughly to test all the features.
1. Save the changes and expect it to get saved and loads correctly after refreshing the page.

#### Ensure the “visual” and “raw” html editor are working fine
1. Run the studio server from devstack dir “make studio-up”
1. Login to studio and create a new course.
1. Add a new unit in the course and go to the unit editor page.
1. Click “HTML” component button and select “Raw”.
1. Click the component to edit, this will show a modal containing “codemirror” editor.
1. Click the “setting-icon” in the editor modal and change the “Editing mode” to “visual”.
1. Click the component to edit, this will show a modal containing “tinyMce” editor.

#### Ensure the “codemirror-plugin” works fine in the tinymce editor
1. Run the studio server from devstack dir “make studio-up”
1. Login to studio and create a new course.
1. Add a new unit in the course and go to the unit editor page.
1. Click “HTML” component button and select “Text”.
1. Click the component to edit, this will show a modal containing “tinymce” editor.
1. Click on “HTML” icon on the toolbar of the editor, this will show a modal with codemirror editor.
1. Add some HTML content and save the changes.
1. The saved changes should show up in the TinyMCE editor.

#### Testing announcements message in studio
1. Run the studio server from devstack dir “make studio-up”
1. Go to “/maintenance” and select “announcement” section.
1. Click “Create new announcement” button.
1. The TinyMCE editor will show up, make sure to test all the components in the editor.
1. Save the announcement message.

## Screenshots and videos of UI changes
**Video comparing TinyMCE4 and TinyMCE5 in studio**

https://user-images.githubusercontent.com/16653571/166076672-72165416-fdc6-49ad-a64c-0a8139fc525b.mp4

#### OLD TinyMCE UI in Studio
![image](https://user-images.githubusercontent.com/16653571/166076360-dc2da1d9-2bcf-4ed7-80d9-406d7d8744b3.png)
#### NEW TinyMCE UI in Studio
![image](https://user-images.githubusercontent.com/16653571/166076334-aa3ce673-1afb-4d31-b7dc-40793d12a551.png)
#### OLD TinyMCE UI on announcement editor page
![image](https://user-images.githubusercontent.com/16653571/166077813-a6f29de0-6ca4-4b8a-975e-0b409c966781.png)
#### NEW TinyMCE UI on announcement editor page
![image](https://user-images.githubusercontent.com/16653571/166077969-96bfbba1-13cc-4c7c-a8ae-1dc97fb8bfab.png)

DubeySandeep committed:
feat: upgrades tinyMCE4 to tinyMCE5

Co-authored-by: Arunmozhi <arunmozhi@opencraft.com>

openedx-webhooks commented:
Thanks for the pull request, @DubeySandeep!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

DubeySandeep committed:
feat: adds new skin studio-tmce5 similar to studio-tmce4

DubeySandeep commented on common/lib/xmodule/xmodule/js/src/html/edit.js:
```diff
@@ -125,19 +125,23 @@
           codemirror: {
-            path: baseUrl + "js/vendor"
+            path: baseUrl + "js/vendor",
+            disableFilesMerge: true,
```
The new codemirror plugin allows passing the custom jsFiles to the plugin. Earlier we had to change the plugin code ([ref](https://github.com/openedx/edx-platform/blob/2e41384102418590a8b6eed7262a0825b542f6e3/common/static/js/vendor/tinymce/js/tinymce/plugins/codemirror/source.html#L85))

DubeySandeep commented on common/lib/xmodule/xmodule/js/src/html/edit.js:
```diff
@@ -99,8 +99,8 @@
-          theme: "modern",
-          skin: 'studio-tmce4',
+          theme: "silver",
+          skin: "studio-tmce5",
```
"studio-tmce5" is the new skin created having similar colour and style of "studio-tmce4" (Note: The custom-skin structure is completely changed in v5 so we had to regenerate the new skin from scratch.)

DubeySandeep commented on common/lib/xmodule/xmodule/js/src/html/edit.js:
```diff
@@ -125,19 +125,23 @@
-            "alignleft aligncenter alignright alignjustify | " +
-            "bullist numlist outdent indent blockquote | link unlink " +
-            ((this.new_image_modal ? 'insertImage' : 'image') + " | code"),
+          toolbar: "formatselect fontselect bold italic underline forecolor wrapAsCode " +
```
Although we have used `|` here but we had disabled the "bars" in the studio-tmce4 skin but disabling the "bars" is not possible in the new skin so ended up remving "|" to achieve the old UI.

Agrendalath replied:
It's also worth noting that these separators are redundant now. The only thing they were doing before, was wrapping the toolbar items in groups. I.e. if a button from a single group should be moved to a lower row (due to insufficient width of the editor), then the whole group is moved there instead.

It is no longer the case because the default [toolbar modes](https://www.tiny.cloud/docs/configure/editor-appearance/#toolbar_mode) have been changed to more user-friendly options. The PC version is using [the floating toolbar](https://www.tiny.cloud/docs/configure/editor-appearance/#floating), but for this to happen, a browser's window width would need to be under 1024px. On [mobile devices](https://www.tiny.cloud/docs/mobile/#mobiledefaultsforselectedsettings), the overflow is handled by [scrolling](https://www.tiny.cloud/docs/configure/editor-appearance/#scrolling).

DubeySandeep commented on common/lib/xmodule/xmodule/js/src/html/edit.js:
```diff
@@ -1233,25 +1237,25 @@
-        image: baseUrl + "images/ico-tinymce-code.png",
-        onclick: function() {
+        tooltip: gettext('Code block'),
+        icon: 'code-sample',
```
TinyMCE5 only allows svg icons and have different way to add new icon ([ref](https://www.tiny.cloud/docs/ui-components/typesoftoolbarbuttons/#configoptions)), overe here we are using a builtin icon.

Old png icon: `{ }`
New svg icon (code-sample): `{ ; }`

(Check screenshot in the PR description for UI)

DubeySandeep reviewed (commented):
(Leaving some inline comments to explain the changes to reviewer.)

natabene commented:
@DubeySandeep Thank you for your contribution. Is this ready for our review?

jmyatt commented:
@jmbowman @jristau1984 as this change affects Studio as opposed to the LMS, maybe it should be reviewed by TNL?

DubeySandeep commented:
@natabene Thanks for checking! Yeah, the PR is ready for review, I've asked @Agrendalath to take the first pass (as an internal reviewer) and I'll be happy to get feedback/suggestion from other reviewers! :)

Agrendalath commented on common/static/js/vendor/tinymce/BUILD_README.txt:
```diff
@@ -1,20 +1,29 @@
-5. Run the following command in the tinymce directory:
-   npx jake minify bundle[themes:*,plugins:*]
-6. Cleanup by deleting the Unversioned files that were created from unzipping jake_package.zip.
+1. Download the desired version from https://github.com/tinymce/tinymce/releases
```
I don't see any releases on this page. There are [tags](https://github.com/tinymce/tinymce/tags), though.

DubeySandeep replied:
That's the old link, I've updated it wit the new link!

tecoholic replied:
This is fixed now.

Agrendalath commented on common/static/js/vendor/tinymce/BUILD_README.txt:
```diff
@@ -1,20 +1,29 @@
+2. If it’s a major update, follow the official migration doc and update the codebase as needed.
+3. Setup the codemirror-plugin as per the instruction below.
+4. Find all the EDX specific changes in the currently used version of tinymce by searching for the string "EDX" in the vendor/js/tinymce dir.
+5. Merge the EDX specific changes with the newly downloaded version.
```
Just for better visibility during the review. We have only two overrides - one in `plugins/codemirror/source.html` and one in `js/tinymce/plugins/codemirror/plugin.js`. Is it correct?

DubeySandeep replied:
Yes, that's true!
![image](https://user-images.githubusercontent.com/16653571/167267869-6f2b6c22-f3d9-4c69-8b7e-846abbdccc73.png)

Agrendalath commented on common/static/js/vendor/tinymce/BUILD_README.txt:
```diff
@@ -1,20 +1,29 @@
+2. Open terminal and change directory to the newly downloaded tinymce.
+3. Run the build command: “yarn build”, this will create multiple zip files in the build directory.
+4. Unzip the tinymce_<version>.zip file in common/static/vendor/js/
+5. Unzip vendor_extra/tinymce/JakePackage.zip in  common/static/vendor/js/tinymce/.
```
I believe it's `common/static/js/vendor/tinymce/`,

Also, is there a specific reason to use `js/vendor/tinymce/js/tinymce` instead of just `js/vendor/tinymce`?

DubeySandeep replied:
> Also, is there a specific reason to use js/vendor/tinymce/js/tinymce instead of just js/vendor/tinymce?

the `common/static/js/vendor/` is the open-edx structure for storing vendor files where as `js/tinymce/` is the tinymce's structure (the js dir contains other files like README, packge.json etc.).

DubeySandeep replied:
> I believe it's common/static/js/vendor/tinymce/,

Yeah, that's true! I've updated the instructions.

Agrendalath replied:
> where as js/tinymce/ is the tinymce's structure

I couldn't find any references to this structure in TinyMCE or CodeMirror, but I see that it's referenced like this in multiple places within the edx-platform, so let's keep it as-is for backward compatibility.

> (the js dir contains other files like README, packge.json etc.).

I don't see other files in [vendor/tinymce/js](https://github.com/openedx/edx-platform/tree/9201eb2029128c2849dfc67eafa89afde83f2f8b/common/static/js/vendor/tinymce/js), but I suppose that keeping [BUILD_README.txt](https://github.com/openedx/edx-platform/blob/9201eb2029128c2849dfc67eafa89afde83f2f8b/common/static/js/vendor/tinymce/BUILD_README.txt) in a separate directory increases its visibility :+1:

Agrendalath reviewed (commented):
@DubeySandeep, I've tested that this is working correctly with both announcements and HTML XBlock. I've checked the HTML XBlock in the Studio unit view, the custom pages view, and the bulk email section of the Instructor Dashboard.

Could you please look into [this failing JS test](https://github.com/openedx/edx-platform/runs/6314020661?check_suite_focus=true#step:12:4931)?

natabene commented:
Owning team will want to have a look at this before merging, since they are actively working in this area.

tecoholic commented:
@Agrendalath This was a doozy. I finally traced it to a race condition in the tests where data was being queried before TinyMCE has finished initializing. I have update the tests, to wait until init is complete before testing. Kindly review.

tecoholic commented on common/lib/xmodule/xmodule/js/src/html/edit.js:
```diff
@@ -1406,7 +1410,7 @@
         visualEditor = this.getVisualEditor();
         raw_content = visualEditor.getContent({
-          format: "raw",
+          format: "text",
```
[TinyMCE 5 uses](https://www.tiny.cloud/docs/api/tinymce/tinymce.editor/#getcontent) the value `text` instead of `raw` for the `format`.

Agrendalath commented on common/lib/xmodule/xmodule/js/spec/html/edit_spec.js:
```diff
@@ -6,26 +6,45 @@ describe('HTMLEditingDescriptor', function() {
-      const { data } = this.descriptor.save();
-      expect(data).toEqual('from visual editor');
+
+      var self = this;
```
@tecoholic, please add a quick explanation of why we're doing this.

tecoholic replied:
@Agrendalath Done.

Agrendalath reviewed (approved):
:+1:

* [x] I tested this: checked that it is working correctly with both announcements and HTML XBlock; checked the HTML XBlock in the Studio unit view, the custom pages view, and the bulk email section of the Instructor Dashboard.
* [x] I read through the code
* [x] I checked for accessibility issues
* [x] Includes documentation
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

Agrendalath commented:
@natabene, this is ready for your review.

Agrendalath reviewed (changes requested):
We found some regressions that cannot be reproduced locally. I'm converting it to a draft for now.

tecoholic committed:
fix: prevent tinymce from breaking on optimised builds

tecoholic commented:
@Agrendalath I have added the commit https://github.com/openedx/edx-platform/pull/30335/commits/49e33090a9a02e29044dbd123626d3e313f33d77 with the fix for editor breaking on Studio. The *minified* version of the TinyMCE doesn't seem to play well with the optimized builds of webpack. Once I switched out the minified to a non-minified version in the Webpack configuration and redid a optimized build, things started to work fine.

Another side-effect I noticed is that the size of the `container.js` file dropped drastically. Using the larger input file actually provides a smaller output file. Something to do with tree-shaking I guess.

#### Before - using tinymce.full.min.js - 2.2MB
![min-js-size](https://user-images.githubusercontent.com/383862/171775193-3b368ca4-0aed-4c78-9a88-46136037d257.png)

#### After - using the tinymce.js - 1.4MB
![non-min-js](https://user-images.githubusercontent.com/383862/171775257-eefc81a7-fe51-4390-9404-d65e1849d188.png)

natabene commented:
@Agrendalath Please let me know if this is ready.

Agrendalath commented:
@tecoholic, I deployed this to the production-like environment (which is using a CDN) and noticed that we have missed the following change: https://github.com/openedx/edx-platform/commit/9c1a42b85236d226db82a0d599a913d17588764f.

We need to check if there is something else missing from here: https://github.com/openedx/edx-platform/pull/3155.

tecoholic committed:
fix: update the CodeMirror plugin to use iframe messaging

This commit implements the iframe postMessage based communication
between TinyMCE and the CodeMirror plugin to get around cross-origin
issues when hosting on CDN.

This is a rewritten set of changes for TinyMCE 5 based on a similar
commit[1] for TinyMCE 4.

https://github.com/openedx/edx-platform/commit/9c1a42b85236d226db82a0d599a913d17588764f

tecoholic commented:
@Agrendalath I have made the necessary changes to get the iframe communication working in TinyMCE in [1ed35b7](https://github.com/openedx/edx-platform/pull/30335/commits/1ed35b79e5ccf70f4f969408e39211fe912acd8d). But I am afraid the work is far from being done.

1. While testing I noticed that the Open Response Assessment block is crashing when the editor is set to WYSIWYG. Tried debugging for almost an hour and for some reason I can't find the source file. Let me know if you have some insights into this.
2. ~While the [reference commit](https://github.com/openedx/edx-platform/commit/9c1a42b85236d226db82a0d599a913d17588764f) removes the `plugin.min.js` file in favour of JS build tool, I am afraid the tools have changed since then and it's no longer automatically done. The [instructions related to this](https://github.com/openedx/edx-platform/pull/30335/files#diff-73d3c1e3cc0e33acdffed2bdbba1a20377d01d4a897b17d19d44459753fcc95bR10) in `BUILD_README.txt` is of no help and I am sure, are completely wrong after the current commit, as it will discard the changes. For the current commit, I manually used [esbuild](https://esbuild.github.io/) to minify the plugin file and added it to the commit.~ - DONE at [8de9e66](https://github.com/openedx/edx-platform/pull/30335/commits/8de9e66404390055f3e2356df718f629aa0cf834)
3. ~The CodeMirror files are compressed into a [single file](https://github.com/openedx/edx-platform/blob/master/common/static/js/vendor/codemirror-compressed.js), but currently the `source.html` file references them individually. So that needs to be changed. This is a small change I will probably get this done first.~ - Done at [eaaccc9](https://github.com/openedx/edx-platform/pull/30335/commits/eaaccc9bb6ea724a14e9f6894cc0a80b368bc313)

Let's move this to draft until we wrap all of this up.

tecoholic committed:
refactor: replace CodeMirror addons files with codemirror-compressed in source.html

tecoholic commented on common/static/js/vendor/tinymce/BUILD_README.txt:
```diff
@@ -1,20 +1,29 @@
+3. Run the build command: “yarn build”, this will create multiple zip files in the build directory.
+4. Unzip the tinymce_<version>.zip file in common/static/vendor/js/
+5. Unzip vendor_extra/tinymce/JakePackage.zip in  common/static/vendor/js/tinymce/.
+6. Run the following command in the tinymce directory: npx jake minify bundle[themes:silver,plugins:*]
```
Tried running this line and it doesn't work. The `JakePackage.zip` contains a `Jakefile` that follows the directory structure of TinyMCE 4.x and is not compatible with the version 5.5.1

tecoholic committed:
docs: update BUILD_README with working instructions for bundling

* This commit updates the instructions for updating TinyMCE with working
  set of steps as the previous instructions using the Jakefile was
  outdated.
* The Jakefile zip bundle in vendor_extras is removed as it is no longer
  required

tecoholic commented:
@Agrendalath I tried following the `BUILD_README.txt` to see minify and bundle the changes made to the codemirror plugin and discovered that the build instructions are not usable. So I have attempted to simplify the process in 3ec66e2cfa9768579c939fa9683d89e67dd038e3. Kindly take a look and review the changes. I tested the final bundle and it works on the Maintenance > Announcements page without issue.

tecoholic committed:
docs: update instructions for compressing code-mirror plugin

tecoholic commented:
@Agrendalath I can't seem to find the source of ORA issue. I have tried devstack's `make dev.static.lms`, I have run `python manage.py lms collecstatic` inside the container, somehow I keep getting the old copy (one pointing to TMCE 4) of the `openassessment-tinymce-editor.*.js` file. Let me know if you know how this can be updates.

![image](https://user-images.githubusercontent.com/383862/172988515-e816568a-b7d5-44f2-a1e0-9f7f5a51bb38.png)

Agrendalath commented:
@tecoholic, ORA is in a [separate repository](https://github.com/openedx/edx-ora2/). This XBlock is reusing LMS TinyMCE libraries, but the theme details are hardcoded [here](https://github.com/openedx/edx-ora2/blob/3834b3ba6c93a99cea738ab0d5d287dabc65e7a7/openassessment/xblock/static/js/src/lms/editors/oa_editor_tinymce.js#L7-L41). There is also [a workaround](https://github.com/openedx/edx-ora2/blob/3834b3ba6c93a99cea738ab0d5d287dabc65e7a7/openassessment/xblock/static/js/src/lms/editors/oa_editor_tinymce.js#L122) for TinyMCE 4, which we've introduced while [removing TinyMCE 5](https://github.com/openedx/edx-ora2/pull/1601) because of the incompatibility with LMS.

Agrendalath commented on common/static/js/vendor/tinymce/BUILD_README.txt:
````diff
@@ -1,20 +1,55 @@
+6. Generate a bundled version of the TinyMCE with all the plugins using the following command
+    ```
+    cd common/static/js/vendor/tinymce/js/tinymce
+    cat tinymce.min.js */*/*.min.js plugins/emoticons/js/emojis.min.js > tinymce.full.min.js
````
_Side note: instead of adding `plugins/emoticons/js/emojis.min.js` separately, we could simply use `*/**/*.min.js`._

I ran it locally and it produced a different result than we have in the repo. I tried generating it with `GNU bash, version 5.1.16(1)-release` and `zsh 5.9`. Isn't this file up-to-date? Also, not all shells are POSIX-compliant (e.g. `fish`), so glob patterns could list files in a different order for some users (thus producing different results). What was wrong with the previous approach (8de9e66404390055f3e2356df718f629aa0cf834)? Was it the missing wildcard before `jquery*`? If I remember correctly, the default Mac OS version of `find` is using full matches.
`find . -name "*.min.js" -not -name "*jquery*" -not -name "tinymce.full.min.js" | sort -r | xargs cat > tinymce.full.min.js`

tecoholic replied:
@Agrendalath I tried the pattern `*/**/*.min.js` in the `cat` it still missed out the `emojis.min.js` file for some reason. So I have kept it as it is.

About using the `find` command, I didn't find any problems with it. I just found a simpler command [here](https://github.com/openedx/edx-platform/pull/3155/commits/9d9b2b8e185f5909abd2abb618aa4d5a7c6fce09) when going through the commits of #3155 and used it as it is much easier to read & understand and does the same task.

Agrendalath replied:
@tecoholic,
> I tried the pattern */**/*.min.js in the cat it still missed out the emojis.min.js file for some reason. So I have kept it as it is.

Ah, right I forgot that the globstar is disabled by default in bash (it needs to be enabled with `shopt -s globstar` to work this way).

> About using the find command, I didn't find any problems with it. I just found a simpler command [here](https://github.com/openedx/edx-platform/pull/3155/commits/9d9b2b8e185f5909abd2abb618aa4d5a7c6fce09) when going through the commits of https://github.com/openedx/edx-platform/pull/3155 and used it as it is much easier to read & understand and does the same task.

I still believe that producing a deterministic (in this case system-agnostic) output is better. We should not assume that people will remember the manual work required when any changes are applied to the TinyMCE library or any of its plugins. If we had the same result on every machine, we could easily verify whether this file has been updated.
I'll let you choose which command you would like to keep, though.

tecoholic replied:
@Agrendalath I have been scratching my head over how this can be done. 
* The globbing pattern in a shell always returns a [consistent list](https://serverfault.com/questions/122737/in-bash-are-wildcard-expansions-guaranteed-to-be-in-order)
* But the output will [vary between systems](https://unix.stackexchange.com/questions/368318/does-the-bash-star-wildcard-always-produce-an-ascending-sorted-list) base on LC_COLLATE, LOCALE values. So consistent between machines cannot be guaranteed.
* The different shells might also implement this differently. I tested with Bash and Fist and got different outputs for `tinymce.full.min.js`

With this in mind I think storing the list of .min.js files explicitly as a file and then using something like `xargs cat < filelist > tinymce.full.min.js`might provide a consistent output irrespective of the environment. I have tested this Bash and Fish and the output are consistent. However this comes with the drawback of keeping an extra file which might not be maintained, if a plugin was added or removed.

What do you think?

Agrendalath replied:
@tecoholic, I'm not sure that I follow… Why don't we simply use `| LC_ALL=C sort -r |`?

tecoholic replied:
@Agrendalath I over-complicated it as I needed to maintain the order as it would load in browser 
1. tinymce
2. plugins (the order of plugins themselves don't matter)
3. plugins' subscripts
and also make sure the glob pattern can ignore jquery.tinymce.min.js and the existing `tinymce.full.min.js` in the folder.

I think the following produces consistent output and I have update the README and the 'tinymce.full.min.js` in 9f139136c85961b2afb1a287bb26828ab19eb773
```
LC_ALL=C cat tinymce.min.js */*/*.min.js plugins/emoticons/js/emojis.min.js > tinymce.full.min.js
```

Agrendalath replied:
@tecoholic, I see, thanks for explaining this. Yep, this solution sounds reasonable 👍🏻

Agrendalath commented on common/static/js/vendor/tinymce/js/tinymce/plugins/codemirror/plugin.js:
```diff
@@ -113,34 +173,45 @@ tinymce.PluginManager.add('codemirror', function(editor, url) {
                     editor.nodeChanged();
                 }
+                editor.undoManager.add();
+                win.close();
```
What is this part doing? I don't see it anywhere in the upstream file.

tecoholic replied:
Removed this.

Agrendalath commented on common/static/js/vendor/tinymce/js/tinymce/plugins/codemirror/source.html:
```diff
@@ -12,12 +12,37 @@
+
+// EDX: Add message handler for listening to events from TinyMCE
+messageHandler = function(event) {
+    var cleanup = function () {
```
We need to [check the origin](https://github.com/openedx/edx-platform/blob/e4de5d2edaee06def4158b49831653272cf7d3da/common/static/js/vendor/tinymce/js/tinymce/plugins/codemirror/source.html#L25-L27).

tecoholic replied:
Check [has been added](https://github.com/openedx/edx-platform/pull/30335/files#diff-55a2764a5d54b5583e51b4de94844fe5e1b0dcd1632f58505ff706ac411fb115R26)

Agrendalath commented on common/static/js/vendor/tinymce/js/tinymce/plugins/codemirror/source.html:
```diff
@@ -30,13 +55,24 @@
-    disableFilesMerge: userSettings.disableFilesMerge || false,
+	    path: codemirrorPath,
+        indentOnInit: false,
+        disableFilesMerge: false,
```
Nit: it seems that we're removing the code for `disableFilesMerge`, so it might not be doing anything anymore.

tecoholic replied:
Removed this.

tecoholic committed:
fix: validate message origin in codemirror source.html

tecoholic committed:
fix: remove unnecssary code in the codemirror plugin

tecoholic committed:
docs: remove outdated note from build instructions

tecoholic commented:
@Agrendalath I rebased this on master and cleaned up the sticking points on this and it should be ready for a final review. Kindly take a look.

Agrendalath commented:
@tecoholic, what about the ORA2 updates I've mentioned [above](https://github.com/openedx/edx-platform/pull/30335#issuecomment-1153885563)? We cannot merge this with a broken XBlock.

Agrendalath commented on common/static/js/vendor/tinymce/BUILD_README.txt:
````diff
@@ -1,20 +1,54 @@
+    cat tinymce.min.js */*/*.min.js plugins/emoticons/js/emojis.min.js > tinymce.full.min.js
+    ```
+
+**NOTE:** Regenerate the `tinymce.full.min.js` bundle everytime the code-mirror `plugin.min.js` is regenerated to ensure the latest changes are added to the bundle.
````
Nit: we're using Markdown syntax here, so we could as well change this file's extension to `md`. This way, it will be rendered in GitHub.

tecoholic commented:
> @tecoholic, what about the ORA2 updates I've mentioned [above](https://github.com/openedx/edx-platform/pull/30335#issuecomment-1153885563)? We cannot merge this with a broken XBlock.

@Agrendalath I have created a PR in ORA (https://github.com/openedx/edx-ora2/pull/1869) to update the dependencies to match this PR.

tecoholic committed:
refactor: rename BUILD_README from txt to md

tecoholic committed:
temp: set the ora version to PR version for testing

tecoholic committed:
refactor: regenerate the tinymce.full.min.js using LC_ALL set to C

Agrendalath reviewed (approved):
:+1:

* [x] I tested this: tested this on [the sandbox](https://bb-6151-sandbox.opencraft.hosting) and locally
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

Agrendalath commented on requirements/edx/base.txt:
```diff
@@ -22,6 +22,8 @@
 -e git+https://github.com/edx-solutions/xblock-google-drive.git@2d176468e33c0713c911b563f8f65f7cf232f5b6#egg=xblock-google-drive
     # via -r requirements/edx/github.in
+-e git+https://github.com/open-craft/edx-ora2.git@tecoholic/BB-6151-upgrade-tinymce-to-v5#egg=ora2
+    # temporary - for testing
```
Leaving a comment to remember that we need to synchronize merging this with https://github.com/openedx/edx-ora2/pull/1869.

tecoholic replied:
@Agrendalath I have updated the ORA2 Versions in 82a5cb3a14

natabene commented:
@Agrendalath Thank you for reviewing, can you see this through to the end?

Agrendalath commented:
@natabene, sure. We will need to merge this shortly after https://github.com/openedx/edx-ora2/pull/1869, so we need one more reviewer. Do you think @pomegranited could review that PR as the (only) CC of `edx-ora2`?

Also, @pomegranited, would you like to review https://github.com/openedx/edx-ora2/pull/1869?

pomegranited commented:
@Agrendalath I'd love to :) Will create a task for next sprint to do this.

And I'm happy to leave a review here too, but we'll still need a CC or someone else with merge rights to merge this.

Agrendalath commented:
@pomegranited,

> And I'm happy to leave a review here too, but we'll still need a CC or someone else with merge rights to merge this.

[I've already reviewed this](https://github.com/openedx/edx-platform/pull/30335#pullrequestreview-1033271660). The "requested changes" are here only to prevent an accidental merge :)

We can coordinate merging these in the sprint after the next one.

pomegranited commented on common/static/js/vendor/tinymce/js/tinymce/jquery.tinymce.min.js:
```diff
@@ -1 +1,92 @@
-!function(e){function t(){function t(e){"remove"===e&&this.each(function(e,t){var n=r(t);n&&n.remove()}),this.find("span.mceEditor,div.mceEditor").each(function(e,t){var n=tinymce.get(t.id.replace(/_parent$/,""));n&&n.remove()})}function i(e){var n,i=this;if(null!=e)t.call(i),i.each(function(t,n){var i;(i=tinymce.get(n.id))&&i.setContent(e)});else if(i.length>0&&(n=tinymce.get(i[0].id)))return n. ...
```
This file is no longer minified, despite its `.min.js` extension. Should it be?

pomegranited replied:
@tecoholic Ah ok, if that's how TinyMCE ships it, then I think it's better not to mess with it. Thanks for [clarifying](https://github.com/openedx/edx-platform/pull/30335#issuecomment-1206402204)!

pomegranited reviewed (approved):
:+1: Works perfectly @tecoholic ! Great work here.

I left [one comment about an un-minified `.min.js` file](https://github.com/openedx/edx-platform/pull/30335#pullrequestreview-1053382414), and there's of course the [dependency update](https://github.com/openedx/edx-platform/pull/30335#pullrequestreview-1033272721) to do before merging. But this change works exactly as described.

- [x] I tested this on my devstack using the PR testing instructions.
- [x] I read through the code
- [x] I checked for accessibility issues by following the [TinyMCE accessibility instructions](https://www.tiny.cloud/docs/advanced/accessibility/).
- [x] Includes documentation on how to upgrade TinyMCE -- thank you for this!
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)

tecoholic commented:
@pomegranited The file is copied from the tinymce build directory IIRC. I think TinyMCE build has the file with code minified but with the comments left in-place. Do you think we should minify it without comments and add the instruction to the BUILD_README.md?

leangseu-edx commented on cms/templates/maintenance/_announcement_edit.html:
```diff
@@ -36,13 +36,15 @@ <h3 class="info-course">
               "insertdatetime media table contextmenu paste"
           ],
-          toolbar: "undo redo | styleselect | bold italic | alignleft aligncenter alignright alignjustify | bullist numlist outdent indent | link image"
+          toolbar: "undo redo styleselect bold italic alignleft aligncenter alignright alignjustify bullist numlist outdent indent link image"
```
Why change this behavior?

Agrendalath replied:
@leangseu-edx, please see [this comment](https://github.com/openedx/edx-platform/pull/30335#discussion_r862248651).

leangseu-edx commented on common/static/js/vendor/tinymce/BUILD_README.md:
````diff
@@ -0,0 +1,53 @@
+6. Install [uglify-js](https://www.npmjs.com/package/uglify-js) and generate `plugin.min.js`
+    ```
+    cd common/static/js/vendor/tinymce/js/plugins/codemirror/
+    uglify plugin.js -m -o plugin.min.js
````
Why do we need to regenerate the plugin? Is it not possible to just drop the code into the `plugin` folder?

Agrendalath replied:
@leangseu-edx, as described in the previous point, `common/static/js/vendor/tinymce/js/tinymce/plugins/codemirror/plugin.js` has custom changes (code drift). We've already ported them to TinyMCE v5, as you can see [here](https://github.com/openedx/edx-platform/pull/30335/files/9f139136c85961b2afb1a287bb26828ab19eb773#diff-83ed1677d3c5797ca90adcc2bb891d16249cec81d187e96b94984f5dec67a73bR23). In this step we are applying these customizations to the minified plugin file.

pomegranited commented:
@tecoholic @Agrendalath I've merged your https://github.com/openedx/edx-ora2/pull/1869 and tagged it: https://github.com/openedx/edx-ora2/tree/4.5.0

Thank you for working to get this change upstream!

tecoholic committed:
chore: update ora2 requirement version to 4.5.0

tecoholic commented:
@pomegranited Thank you :)

@Agrendalath I have removed the ora2's pin from custom branches and set the version to 4.5.0. There seems to be some merge conflicts in the requirements files. Do you want me to rebase on master and update the PR?

Agrendalath commented:
@tecoholic, yes, please rebase it.

tecoholic commented:
@Agrendalath Rebased on master and resolved conflicts.

Agrendalath reviewed (approved):
All good from my end. Linking my review: https://github.com/openedx/edx-platform/pull/30335#pullrequestreview-1033271660

openedx-webhooks commented:
@DubeySandeep 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/30335&entry.1484655318=https://github.com/DubeySandeep&entry.752974735=2022-04-29+21:28&entry.1917517419=2022-09-19+10:43&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
